### ISSUE TITLE
ProxyManagerBridge should possibly use Lazy Loading Ghost Object Proxies instead of Lazy Loading Value Holder Proxies

### ISSUE THREAD

**Issue by ostrolucky** [symfony/symfony#35345]
I have created new issue to make it less doctrine specific.

Here is demonstration of different lazy proxies (modified [proxy-manager examples]( https://github.com/Ocramius/ProxyManager/tree/master/examples))

```php

use ProxyManager\Factory\LazyLoadingGhostFactory;
use ProxyManager\Factory\LazyLoadingValueHolderFactory;

class Foo
{
    public function compare(self $foo)
    {
        var_dump($foo === $this);
    }
}


$proxy = (new LazyLoadingValueHolderFactory())->createProxy(Foo::class, function (&$wrappedObject) {
    $wrappedObject = new Foo();
    return true;
});

$proxy->compare($proxy); // bool(false)

$proxy = (new LazyLoadingGhostFactory())->createProxy(Foo::class, fn() => true);
$proxy->compare($proxy); // bool(true)
```

First approach is used by Symfony. As you can see, it breaks identity and is more complicated to use. I suggest to use second method, which doesn't break identities. Would solve https://github.com/symfony/symfony/issues/35216 and issues related to it.

cc @stof

**Comment by nicolas-grekas** [comment]
I didn't look closely enough to both type of proxies to anticipate what kind of behaviors this could change, potentially breaking BC. Would you mind investigating, and working on a patch maybe?

**Comment by ostrolucky** [comment]
Nope, this part of symfony codebase is too unreadable to me.

**Comment by nicolas-grekas** [comment]
Looking at https://github.com/Ocramius/ProxyManager/blob/master/docs/lazy-loading-ghost-object.md this isn't a fit: ghost objects require writting initializers that bypass constructors and know how to define properties. Thanks for considering.

**Comment by ostrolucky** [comment]
Looks like we can do 
```php
use ProxyManager\Factory\LazyLoadingGhostFactory;
use ProxyManager\Proxy\GhostObjectInterface;

class Foo
{
    /**
     * @var string
     */
    private $foo;

    public function __construct(string $foo)
    {
        $this->foo = $foo;
    }

    public function compare(self $foo)
    {
        var_dump($foo === $this);
    }

    public function getFoo()
    {
        return $this->foo;
    }
}

$proxy = (new LazyLoadingGhostFactory())->createProxy(Foo::class, fn(GhostObjectInterface $proxy) => $proxy->__construct('yeye'));
$proxy->compare($proxy); // bool(true)
var_dump($proxy->getFoo($proxy)); // string(4) "yeye"
```

:)

**Comment by Ocramius** [comment]
Yeah, calling the constructor in a delayed fashion is possible, just annoying to do, and incompatible with creating lazy types whenever a factory is specified for a service.

Calling setters & initializers should also be feasible.

In practice, this proxy type is much safer to use, since `spl_object_id($proxy) === spl_object_id($realInstance)`, which isn't true for the default implementation (which is arguably simpler to understand)

**Comment by stof** [comment]
The complexity of using a ghost proxy is that it requires a different way to initialize services than when using non-proxy services. For the value object proxy, the bridge can use the existing non-lazy instantiation for the initialization, meaning that we almost never need to modify the bridge when adding new features in the DI component and that we keep the complexity low.

**Comment by nicolas-grekas** [comment]
Ghosts aren't compatible with immutable setters, isn't it?

**Comment by stof** [comment]
probably not, indeed.

**Comment by stof** [comment]
and not with factory methods either.

**Comment by Ocramius** [comment]
@nicolas-grekas yes, they are compatible AFAIK.

**Comment by nicolas-grekas** [comment]
With immutable setters and with factories, the process would be to create a separate instance, then copy all its state to the properties of the proxy instance, isn't it? Looks fragile to me, but I may be wrong. Looks quite some work also (and I should be less wrong here :) )

**Comment by nicolas-grekas** [comment]
I just had a second look at this: ghost objects indeed would work with immutable setters and factories.

They do have one requirement: the exact class of the service must be known at compilation time.

This doesn't work with factories that return a different type of object based on its argument (typically: returning a specific cache adapter based on a DSN). And it doesn't work either with interface-based proxies.

I suppose this can be solved by adding a rule that if a service definition declares an abstract class, then we go with a lazy value holder (aka a virtual proxy) and if the class is a concrete one, then we go with the ghost object.

I'd be happy to review a PR that tries to do so.

**Comment by nicolas-grekas** [comment]
https://github.com/Ocramius/ProxyManager/issues/760

**Comment by nicolas-grekas** [comment]
While using a ghost-object proxy would be nice for the generic and default case, it cannot be used for entity managers until https://github.com/doctrine/orm/issues/5933 is solved. The reason is that while a bit hacky, using a value-holder proxy is currently the only way to reset the entity manager that works. :shrug: 
See also https://github.com/doctrine/DoctrineBundle/pull/1529

**Comment by Ocramius** [comment]
> The reason is that while a bit hacky, using a value-holder proxy is currently the only way to reset the entity manager that works. shrug

Resetting the whole state of the ORM is also feasible (calling `__construct()` again?).

To be noted that none of this will work, once `readonly` starts to become more popular :scream:

**Comment by nicolas-grekas** [comment]
I implemented this in #46458, including the reset by calling __construct for entity managers. I'd be happy to have reviews there :pray:
